app/code/Bogdank/SupportChat/Observer/ChatUserLogin.php

```diff
@@ -20,29 +21,50 @@ class ChatUserLogin implements ObserverInterface
      */
     private $messageCollectionFactory;
 
+    /**
+     * @var \Magento\Framework\DB\TransactionFactory $transactionFactory
+     */
+    private $transactionFactory;
+
     /**
      * ChatUserLogin constructor.
-     * @param GenerateHash $generateHash
+     * @param ChatHashManager $generateHash
      * @param \Bogdank\SupportChat\Model\ResourceModel\SupportMessage\CollectionFactory $messageCollectionFactory
+     * @param \Magento\Framework\DB\TransactionFactory $transactionFactory
      */
     public function __construct(
-        \Bogdank\SupportChat\Model\GenerateHash $generateHash,
-        \Bogdank\SupportChat\Model\ResourceModel\SupportMessage\CollectionFactory $messageCollectionFactory
-    ) {
+        \Bogdank\SupportChat\Model\ChatHashManager $generateHash,
+        \Bogdank\SupportChat\Model\ResourceModel\SupportMessage\CollectionFactory $messageCollectionFactory,
+        \Magento\Framework\DB\TransactionFactory $transactionFactory
+    )
+    {
         $this->generateHash = $generateHash;
         $this->messageCollectionFactory = $messageCollectionFactory;
+        $this->transactionFactory = $transactionFactory;
     }
+
     public function execute(Observer $observer)
     {
         /** @var \Magento\Customer\Model\Data\Customer $customer */
         $customer = $observer->getData('customer');
-        $this->updateChatMessages($customer);
+        $this->updateChatMessagesData($customer);
     }
 
-    private function updateChatMessages(): void
+    /**
+     * @param \Magento\Customer\Model\Data\Customer $customer
+     * @throws \Exception
+     */
+    public function updateChatMessagesData($customer): void
     {
-        /** @var \Bogdank\SupportChat\Model\ResourceModel\SupportMessage\CollectionFactory $messageCollectionFactory */
         $chatCollection = $this->messageCollectionFactory->create();
-        $chatCollection->addFieldToFilter('chat_hash', $this->generateHash->getChatHashCookie());
+        /** @var \Magento\Framework\DB\Transaction $transaction */
+        $transaction = $this->transactionFactory->create();
+        $chatCollection->addFieldToFilter('chat_hash', $this->generateHash->getChatHash());
+        /** @var \Bogdank\SupportChat\Model\SupportMessage $supportMessage */
+        foreach ($chatCollection as $supportMessage) {
+            $supportMessage->setUserId((int)$customer->getId());
+            $transaction->addObject($supportMessage);
+            $transaction->save();
```

Comment:
@holmatro90 Data must be saved after the loop! This decreases the performance and the chance of fail, not increases it